File: src/sdk/cpp/ofa_agent/src/offline.cpp
/**
 * @file offline.cpp
 * @brief Offline Support Implementation
 * Sprint 30: Offline Capability Enhancement
 */

#include "ofa/offline.hpp"
#include <chrono>
#include <thread>
#include <random>
#include <sstream>
#include <fstream>
#include <algorithm>

namespace ofa {

// Helper: generate UUID
static std::string generateId() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);

    std::stringstream ss;
    ss << std::hex;
    for (int i = 0; i < 8; i++) ss << dis(gen);
    ss << "-";
    for (int i = 0; i < 4; i++) ss << dis(gen);
    return ss.str();
}

// Helper: current timestamp
static int64_t nowMs() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();
}

// ==================== LocalScheduler ====================

LocalScheduler::LocalScheduler(size_t maxConcurrent)
    : maxConcurrent_(maxConcurrent) {
}

LocalScheduler::~LocalScheduler() {
    stop();
}

void LocalScheduler::registerHandler(const std::string& skillId, TaskHandler handler) {
    std::lock_guard<std::mutex> lock(mutex_);
    handlers_[skillId] = handler;
}

std::string LocalScheduler::submit(const std::string& skillId,
                                    const std::string& operation,
                                    const json& input,
                                    int maxRetries) {
    std::lock_guard<std::mutex> lock(mutex_);

    LocalTask task;
    task.taskId = "local-" + generateId();
    task.skillId = skillId;
    task.operation = operation;
    task.input = input;
    task.status = LocalTaskStatus::Pending;
    task.maxRetries = maxRetries;
    task.createdAt = nowMs();
    task.updatedAt = nowMs();
    task.syncPending = true;

    pendingQueue_.push(task);
    stats_.totalTasks++;

    cv_.notify_one();
    return task.taskId;
}

bool LocalScheduler::cancel(const std::string& taskId) {
    std::lock_guard<std::mutex> lock(mutex_);

    // Check pending queue
    std::queue<LocalTask> newQueue;
    bool found = false;
    while (!pendingQueue_.empty()) {
        auto task = pendingQueue_.front();
        pendingQueue_.pop();
        if (task.taskId == taskId) {
            task.status = LocalTaskStatus::Cancelled;
            task.updatedAt = nowMs();
            completedTasks_[taskId] = task;
            found = true;
        } else {
            newQueue.push(task);
        }
    }
    pendingQueue_ = newQueue;

    // Check active tasks
    auto it = activeTasks_.find(taskId);
    if (it != activeTasks_.end()) {
        it->second.status = LocalTaskStatus::Cancelled;
        return true;
    }

    return found;
}

LocalTask LocalScheduler::getTask(const std::string& taskId) const {
    std::lock_guard<std::mutex> lock(mutex_);

    auto it = activeTasks_.find(taskId);
    if (it != activeTasks_.end()) {
        return it->second;
    }

    auto it2 = completedTasks_.find(taskId);
    if (it2 != completedTasks_.end()) {
        return it2->second;
    }

    throw OfAException("Task not found: " + taskId);
}

std::vector<LocalTask> LocalScheduler::getPendingTasks() const {
    std::lock_guard<std::mutex> lock(mutex_);

    std::vector<LocalTask> tasks;
    std::queue<LocalTask> temp = pendingQueue_;
    while (!temp.empty()) {
        tasks.push_back(temp.front());
        temp.pop();
    }
    return tasks;
}

std::vector<LocalTask> LocalScheduler::getSyncPendingTasks() const {
    std::lock_guard<std::mutex> lock(mutex_);

    std::vector<LocalTask> tasks;
    for (const auto& pair : completedTasks_) {
        if (pair.second.syncPending) {
            tasks.push_back(pair.second);
        }
    }
    return tasks;
}

void LocalScheduler::onComplete(TaskCallback callback) {
    onComplete_ = callback;
}

void LocalScheduler::start() {
    if (running_) return;
    running_ = true;
    workerThread_ = std::thread(&LocalScheduler::processLoop, this);
}

void LocalScheduler::stop() {
    if (!running_) return;
    running_ = false;
    cv_.notify_all();
    if (workerThread_.joinable()) {
        workerThread_.join();
    }
}

LocalScheduler::Stats LocalScheduler::stats() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return stats_;
}

void LocalScheduler::processLoop() {
    while (running_) {
        LocalTask task;
        {
            std::unique_lock<std::mutex> lock(mutex_);
            cv_.wait_for(lock, std::chrono::milliseconds(100), [this]() {
                return !pendingQueue_.empty() || !running_;
            });

            if (!running_) break;

            if (pendingQueue_.empty()) continue;

            // Check concurrency limit
            if (activeTasks_.size() >= maxConcurrent_) continue;

            task = pendingQueue_.front();
            pendingQueue_.pop();
            activeTasks_[task.taskId] = task;
        }

        executeTask(task);
    }
}

void LocalScheduler::executeTask(LocalTask& task) {
    task.status = LocalTaskStatus::Running;
    task.updatedAt = nowMs();

    try {
        auto it = handlers_.find(task.skillId);
        if (it == handlers_.end()) {
            throw OfAException("No handler for skill: " + task.skillId);
        }

        task.result = it->second(task.skillId, task.operation, task.input);
        task.status = LocalTaskStatus::Success;
        task.completedAt = nowMs();
        stats_.successTasks++;

        {
            std::lock_guard<std::mutex> lock(mutex_);
            activeTasks_.erase(task.taskId);
            completedTasks_[task.taskId] = task;
        }

        if (onComplete_) {
            onComplete_(task);
        }

    } catch (const std::exception& e) {
        task.error = e.what();
        task.updatedAt = nowMs();

        if (task.retryCount < task.maxRetries) {
            retryTask(task);
        } else {
            task.status = LocalTaskStatus::Failed;
            task.completedAt = nowMs();
            stats_.failedTasks++;

            {
                std::lock_guard<std::mutex> lock(mutex_);
                activeTasks_.erase(task.taskId);
                completedTasks_[task.taskId] = task;
            }

            if (onComplete_) {
                onComplete_(task);
            }
        }
    }

    cv_.notify_one();
}

void LocalScheduler::retryTask(LocalTask& task) {
    task.status = LocalTaskStatus::Retrying;
    task.retryCount++;
    stats_.retryTasks++;
    task.updatedAt = nowMs();

    // Add back to queue with delay
    std::this_thread::sleep_for(std::chrono::seconds(1));

    {
        std::lock_guard<std::mutex> lock(mutex_);
        activeTasks_.erase(task.taskId);
        pendingQueue_.push(task);
    }

    cv_.notify_one();
}

// ==================== OfflineCache ====================

OfflineCache::OfflineCache(const std::string& dbPath)
    : dbPath_(dbPath) {
    if (!dbPath_.empty()) {
        load();
    }
}

OfflineCache::~OfflineCache() {
    if (!dbPath_.empty()) {
        persist();
    }
}

void OfflineCache::put(const std::string& key,
                        const json& data,
                        int64_t ttlSeconds,
                        bool syncPending) {
    std::lock_guard<std::mutex> lock(mutex_);

    CacheEntry entry;
    entry.key = key;
    entry.data = data;
    entry.createdAt = nowMs();
    entry.expiresAt = ttlSeconds > 0 ? entry.createdAt + ttlSeconds * 1000 : 0;
    entry.syncPending = syncPending;
    entry.source = "local";

    cache_[key] = entry;
}

json OfflineCache::get(const std::string& key) const {
    std::lock_guard<std::mutex> lock(mutex_);

    auto it = cache_.find(key);
    if (it == cache_.end()) {
        return json::object();
    }

    if (it->second.isExpired()) {
        return json::object();
    }

    return it->second.data;
}

bool OfflineCache::exists(const std::string& key) const {
    std::lock_guard<std::mutex> lock(mutex_);

    auto it = cache_.find(key);
    if (it == cache_.end()) return false;

    return !it->second.isExpired();
}

void OfflineCache::remove(const std::string& key) {
    std::lock_guard<std::mutex> lock(mutex_);
    cache_.erase(key);
}

void OfflineCache::clearExpired() {
    std::lock_guard<std::mutex> lock(mutex_);

    auto now = std::chrono::duration_cast<std::chrono::seconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();

    for (auto it = cache_.begin(); it != cache_.end(); ) {
        if (it->second.isExpired()) {
            it = cache_.erase(it);
        } else {
            ++it;
        }
    }
}

std::vector<CacheEntry> OfflineCache::getSyncPending() const {
    std::lock_guard<std::mutex> lock(mutex_);

    std::vector<CacheEntry> entries;
    for (const auto& pair : cache_) {
        if (pair.second.syncPending && !pair.second.isExpired()) {
            entries.push_back(pair.second);
        }
    }
    return entries;
}

void OfflineCache::markSynced(const std::string& key) {
    std::lock_guard<std::mutex> lock(mutex_);

    auto it = cache_.find(key);
    if (it != cache_.end()) {
        it->second.syncPending = false;
    }
}

size_t OfflineCache::size() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return cache_.size();
}

void OfflineCache::persist() {
    if (dbPath_.empty()) return;

    std::lock_guard<std::mutex> lock(mutex_);

    json j = json::array();
    for (const auto& pair : cache_) {
        json entry;
        entry["key"] = pair.second.key;
        entry["data"] = pair.second.data;
        entry["createdAt"] = pair.second.createdAt;
        entry["expiresAt"] = pair.second.expiresAt;
        entry["syncPending"] = pair.second.syncPending;
        entry["source"] = pair.second.source;
        j.push_back(entry);
    }

    std::ofstream file(dbPath_);
    if (file.is_open()) {
        file << j.dump(2);
        file.close();
    }
}

void OfflineCache::load() {
    if (dbPath_.empty()) return;

    std::lock_guard<std::mutex> lock(mutex_);

    std::ifstream file(dbPath_);
    if (!file.is_open()) return;

    try {
        json j;
        file >> j;
        file.close();

        for (const auto& entry : j) {
            CacheEntry ce;
            ce.key = entry["key"].get<std::string>();
            ce.data = entry["data"];
            ce.createdAt = entry["createdAt"].get<int64_t>();
            ce.expiresAt = entry["expiresAt"].get<int64_t>();
            ce.syncPending = entry["syncPending"].get<bool>();
            ce.source = entry["source"].get<std::string>();

            if (!ce.isExpired()) {
                cache_[ce.key] = ce;
            }
        }
    } catch (...) {
        // Invalid JSON, ignore
    }
}

// ==================== OfflineManager ====================

OfflineManager::OfflineManager(const std::string& cachePath)
    : cache_(cachePath) {
}

OfflineManager::~OfflineManager() {
    stop();
}

void OfflineManager::setLevel(OfflineLevel level) {
    level_ = level;
}

bool OfflineManager::isOnline() const {
    // Simple check: try to resolve a known host
    // In production, implement proper network detection
    return level_ == OfflineLevel::L4_Online;
}

void OfflineManager::autoSwitch() {
    // Auto-detect network and switch mode
    // For now, keep current level
    // In production: ping center, check connectivity
}

LocalScheduler::TaskHandler OfflineManager::submitTask(const std::string& skillId,
                                                         const std::string& operation,
                                                         const json& input) {
    // Return empty handler, actual submit happens in scheduler
    std::string taskId = scheduler_.submit(skillId, operation, input);

    // Cache the task input
    cache_.put("task:" + taskId, input, 0, true);

    return [](const std::string&, const std::string&, const json&) { return json::object(); };
}

std::future<bool> OfflineManager::sync() {
    return std::async(std::launch::async, [this]() {
        // Sync pending tasks and cache to center
        auto pendingTasks = scheduler_.getSyncPendingTasks();
        auto pendingCache = cache_.getSyncPending();

        // In production: send to center via gRPC/REST
        // For now, just mark as synced

        for (const auto& task : pendingTasks) {
            // Would send to center here
            // Mark as synced
        }

        for (const auto& entry : pendingCache) {
            cache_.markSynced(entry.key);
        }

        return true;
    });
}

void OfflineManager::registerOfflineSkill(const std::string& skillId,
                                           LocalScheduler::TaskHandler handler) {
    scheduler_.registerHandler(skillId, handler);
}

void OfflineManager::start() {
    if (running_) return;
    running_ = true;
    scheduler_.start();
    monitorThread_ = std::thread(&OfflineManager::networkMonitor, this);
}

void OfflineManager::stop() {
    if (!running_) return;
    running_ = false;
    scheduler_.stop();
    if (monitorThread_.joinable()) {
        monitorThread_.join();
    }
    cache_.persist();
}

void OfflineManager::networkMonitor() {
    while (running_) {
        std::this_thread::sleep_for(std::chrono::seconds(5));

        autoSwitch();

        // If online, try to sync
        if (level_ == OfflineLevel::L4_Online) {
            sync().get();
        }
    }
}

} // namespace ofa

File: src/sdk/go/offline.go
package ofa

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// OfflineLevel defines offline capability level
type OfflineLevel int

const (
	OfflineLevelL1 OfflineLevel = iota // Complete offline
	OfflineLevelL2                     // LAN collaboration
	OfflineLevelL3                     // Weak network sync
	OfflineLevelL4                     // Online mode
)

// LocalTaskStatus defines local task status
type LocalTaskStatus int

const (
	TaskStatusPending LocalTaskStatus = iota
	TaskStatusRunning
	TaskStatusSuccess
	TaskStatusFailed
	TaskStatusCancelled
	TaskStatusRetrying
)

// LocalTask represents a local task
type LocalTask struct {
	TaskID      string          `json:"taskId"`
	SkillID     string          `json:"skillId"`
	Operation   string          `json:"operation"`
	Input       json.RawMessage `json:"input"`
	Result      json.RawMessage `json:"result,omitempty"`
	Status      LocalTaskStatus `json:"status"`
	Error       string          `json:"error,omitempty"`
	RetryCount  int             `json:"retryCount"`
	MaxRetries  int             `json:"maxRetries"`
	CreatedAt   int64           `json:"createdAt"`
	UpdatedAt   int64           `json:"updatedAt"`
	CompletedAt int64           `json:"completedAt,omitempty"`
	SyncPending bool            `json:"syncPending"`
}

// CacheEntry represents cache entry
type CacheEntry struct {
	Key         string          `json:"key"`
	Data        json.RawMessage `json:"data"`
	CreatedAt   int64           `json:"createdAt"`
	ExpiresAt   int64           `json:"expiresAt"`
	SyncPending bool            `json:"syncPending"`
	Source      string          `json:"source"`
}

// LocalScheduler manages local task execution
type LocalScheduler struct {
	maxConcurrent int
	handlers      map[string]SkillHandler
	pendingQueue  []*LocalTask
	activeTasks   map[string]*LocalTask
	completedTasks map[string]*LocalTask
	onComplete    func(*LocalTask)

	mu     sync.Mutex
	cond   *sync.Cond
	running bool
	done   chan struct{}

	stats SchedulerStats
}

// SchedulerStats holds scheduler statistics
type SchedulerStats struct {
	TotalTasks   uint64
	SuccessTasks uint64
	FailedTasks  uint64
	RetryTasks   uint64
}

// NewLocalScheduler creates a new local scheduler
func NewLocalScheduler(maxConcurrent int) *LocalScheduler {
	s := &LocalScheduler{
		maxConcurrent:  maxConcurrent,
		handlers:       make(map[string]SkillHandler),
		activeTasks:    make(map[string]*LocalTask),
		completedTasks: make(map[string]*LocalTask),
		done:           make(chan struct{}),
	}
	s.cond = sync.NewCond(&s.mu)
	return s
}

// RegisterHandler registers a task handler
func (s *LocalScheduler) RegisterHandler(skillID string, handler SkillHandler) {
	s.mu.Lock()
	s.handlers[skillID] = handler
	s.mu.Unlock()
}

// Submit submits a new task
func (s *LocalScheduler) Submit(skillID, operation string, input json.RawMessage, maxRetries int) string {
	s.mu.Lock()
	defer s.mu.Unlock()

	task := &LocalTask{
		TaskID:      generateID(),
		SkillID:     skillID,
		Operation:   operation,
		Input:       input,
		Status:      TaskStatusPending,
		MaxRetries:  maxRetries,
		CreatedAt:   time.Now().UnixMilli(),
		UpdatedAt:   time.Now().UnixMilli(),
		SyncPending: true,
	}

	s.pendingQueue = append(s.pendingQueue, task)
	s.stats.TotalTasks++
	s.cond.Signal()

	return task.TaskID
}

// Cancel cancels a task
func (s *LocalScheduler) Cancel(taskID string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Check pending queue
	for i, task := range s.pendingQueue {
		if task.TaskID == taskID {
			task.Status = TaskStatusCancelled
			task.UpdatedAt = time.Now().UnixMilli()
			s.completedTasks[taskID] = task
			s.pendingQueue = append(s.pendingQueue[:i], s.pendingQueue[i+1:]...)
			return true
		}
	}

	// Check active tasks
	if task, ok := s.activeTasks[taskID]; ok {
		task.Status = TaskStatusCancelled
		return true
	}

	return false
}

// GetTask gets task by ID
func (s *LocalScheduler) GetTask(taskID string) (*LocalTask, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if task, ok := s.activeTasks[taskID]; ok {
		return task, nil
	}

	if task, ok := s.completedTasks[taskID]; ok {
		return task, nil
	}

	return nil, fmt.Errorf("task not found: %s", taskID)
}

// GetSyncPendingTasks returns tasks pending sync
func (s *LocalScheduler) GetSyncPendingTasks() []*LocalTask {
	s.mu.Lock()
	defer s.mu.Unlock()

	var tasks []*LocalTask
	for _, task := range s.completedTasks {
		if task.SyncPending {
			tasks = append(tasks, task)
		}
	}
	return tasks
}

// OnComplete sets completion callback
func (s *LocalScheduler) OnComplete(callback func(*LocalTask)) {
	s.mu.Lock()
	s.onComplete = callback
	s.mu.Unlock()
}

// Start starts the scheduler
func (s *LocalScheduler) Start() {
	s.mu.Lock()
	if s.running {
		s.mu.Unlock()
		return
	}
	s.running = true
	s.mu.Unlock()

	go s.processLoop()
}

// Stop stops the scheduler
func (s *LocalScheduler) Stop() {
	s.mu.Lock()
	s.running = false
	s.mu.Unlock()
	s.cond.Broadcast()
	close(s.done)
}

// Stats returns scheduler statistics
func (s *LocalScheduler) Stats() SchedulerStats {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.stats
}

func (s *LocalScheduler) processLoop() {
	for {
		s.mu.Lock()

		// Wait for tasks or stop signal
		for len(s.pendingQueue) == 0 && s.running {
			s.cond.Wait()
		}

		if !s.running {
			s.mu.Unlock()
			return
		}

		// Check concurrency limit
		if len(s.activeTasks) >= s.maxConcurrent {
			s.mu.Unlock()
			time.Sleep(100 * time.Millisecond)
			continue
		}

		// Get next task
		task := s.pendingQueue[0]
		s.pendingQueue = s.pendingQueue[1:]
		s.activeTasks[task.TaskID] = task
		s.mu.Unlock()

		s.executeTask(task)
	}
}

func (s *LocalScheduler) executeTask(task *LocalTask) {
	task.Status = TaskStatusRunning
	task.UpdatedAt = time.Now().UnixMilli()

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	handler, ok := s.handlers[task.SkillID]
	if !ok {
		s.completeTask(task, nil, fmt.Errorf("no handler for skill: %s", task.SkillID))
		return
	}

	output, err := handler(ctx, task.Input)
	s.completeTask(task, output, err)
}

func (s *LocalScheduler) completeTask(task *LocalTask, output []byte, err error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if err != nil {
		task.Error = err.Error()
		task.UpdatedAt = time.Now().UnixMilli()

		if task.RetryCount < task.MaxRetries {
			task.Status = TaskStatusRetrying
			task.RetryCount++
			s.stats.RetryTasks++
			s.activeTasks[task.TaskID] = task // Keep active for retry
			go func() {
				time.Sleep(time.Second)
				s.mu.Lock()
				delete(s.activeTasks, task.TaskID)
				s.pendingQueue = append(s.pendingQueue, task)
				s.cond.Signal()
				s.mu.Unlock()
			}()
			return
		}

		task.Status = TaskStatusFailed
		task.CompletedAt = time.Now().UnixMilli()
		s.stats.FailedTasks++
	} else {
		task.Result = output
		task.Status = TaskStatusSuccess
		task.CompletedAt = time.Now().UnixMilli()
		s.stats.SuccessTasks++
	}

	delete(s.activeTasks, task.TaskID)
	s.completedTasks[task.TaskID] = task

	if s.onComplete != nil {
		go s.onComplete(task)
	}

	s.cond.Signal()
}

// OfflineCache manages offline data cache
type OfflineCache struct {
	dbPath string
	cache  map[string]*CacheEntry
	mu     sync.Mutex
}

// NewOfflineCache creates a new offline cache
func NewOfflineCache(dbPath string) *OfflineCache {
	c := &OfflineCache{
		dbPath: dbPath,
		cache:  make(map[string]*CacheEntry),
	}
	c.load()
	return c
}

// Put stores data in cache
func (c *OfflineCache) Put(key string, data json.RawMessage, ttlSeconds int64, syncPending bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	now := time.Now().UnixMilli()
	entry := &CacheEntry{
		Key:         key,
		Data:        data,
		CreatedAt:   now,
		ExpiresAt:   ttlSeconds > 0 ? now + ttlSeconds*1000 : 0,
		SyncPending: syncPending,
		Source:      "local",
	}
	c.cache[key] = entry
}

// Get retrieves data from cache
func (c *OfflineCache) Get(key string) (json.RawMessage, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	entry, ok := c.cache[key]
	if !ok {
		return nil, false
	}

	if entry.ExpiresAt > 0 && time.Now().UnixMilli() > entry.ExpiresAt {
		return nil, false
	}

	return entry.Data, true
}

// Remove removes data from cache
func (c *OfflineCache) Remove(key string) {
	c.mu.Lock()
	delete(c.cache, key)
	c.mu.Unlock()
}

// ClearExpired clears expired entries
func (c *OfflineCache) ClearExpired() {
	c.mu.Lock()
	defer c.mu.Unlock()

	now := time.Now().UnixMilli()
	for key, entry := range c.cache {
		if entry.ExpiresAt > 0 && now > entry.ExpiresAt {
			delete(c.cache, key)
		}
	}
}

// GetSyncPending returns entries pending sync
func (c *OfflineCache) GetSyncPending() []*CacheEntry {
	c.mu.Lock()
	defer c.mu.Unlock()

	var entries []*CacheEntry
	now := time.Now().UnixMilli()
	for _, entry := range c.cache {
		if entry.SyncPending && (entry.ExpiresAt == 0 || now <= entry.ExpiresAt) {
			entries = append(entries, entry)
		}
	}
	return entries
}

// MarkSynced marks entry as synced
func (c *OfflineCache) MarkSynced(key string) {
	c.mu.Lock()
	if entry, ok := c.cache[key]; ok {
		entry.SyncPending = false
	}
	c.mu.Unlock()
}

// Persist saves cache to disk
func (c *OfflineCache) Persist() {
	if c.dbPath == "" {
		return
	}

	c.mu.Lock()
	defer c.mu.Unlock()

	// Ensure directory exists
	dir := filepath.Dir(c.dbPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		log.Printf("Failed to create cache directory: %v", err)
		return
	}

	// Serialize cache
	var entries []*CacheEntry
	for _, entry := range c.cache {
		if entry.ExpiresAt == 0 || time.Now().UnixMilli() <= entry.ExpiresAt {
			entries = append(entries, entry)
		}
	}

	data, err := json.Marshal(entries)
	if err != nil {
		log.Printf("Failed to marshal cache: %v", err)
		return
	}

	if err := os.WriteFile(c.dbPath, data, 0644); err != nil {
		log.Printf("Failed to write cache: %v", err)
	}
}

func (c *OfflineCache) load() {
	if c.dbPath == "" {
		return
	}

	data, err := os.ReadFile(c.dbPath)
	if err != nil {
		if !os.IsNotExist(err) {
			log.Printf("Failed to read cache: %v", err)
		}
		return
	}

	var entries []*CacheEntry
	if err := json.Unmarshal(data, &entries); err != nil {
		log.Printf("Failed to unmarshal cache: %v", err)
		return
	}

	now := time.Now().UnixMilli()
	for _, entry := range entries {
		if entry.ExpiresAt == 0 || now <= entry.ExpiresAt {
			c.cache[entry.Key] = entry
		}
	}
}

// OfflineManager manages offline operations
type OfflineManager struct {
	level    OfflineLevel
	scheduler *LocalScheduler
	cache    *OfflineCache
	running  bool
	done     chan struct{}
	mu       sync.Mutex
}

// NewOfflineManager creates a new offline manager
func NewOfflineManager(cachePath string) *OfflineManager {
	return &OfflineManager{
		level:     OfflineLevelL4,
		scheduler: NewLocalScheduler(4),
		cache:     NewOfflineCache(cachePath),
		done:      make(chan struct{}),
	}
}

// SetLevel sets offline level
func (m *OfflineManager) SetLevel(level OfflineLevel) {
	m.mu.Lock()
	m.level = level
	m.mu.Unlock()
}

// Level returns current offline level
func (m *OfflineManager) Level() OfflineLevel {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.level
}

// IsOnline checks if online mode
func (m *OfflineManager) IsOnline() bool {
	return m.Level() == OfflineLevelL4
}

// Scheduler returns the scheduler
func (m *OfflineManager) Scheduler() *LocalScheduler {
	return m.scheduler
}

// Cache returns the cache
func (m *OfflineManager) Cache() *OfflineCache {
	return m.cache
}

// SubmitTask submits a local task
func (m *OfflineManager) SubmitTask(skillID, operation string, input json.RawMessage) string {
	taskID := m.scheduler.Submit(skillID, operation, input, 3)

	// Cache the input
	m.cache.Put("task:"+taskID, input, 0, true)

	return taskID
}

// RegisterOfflineSkill registers an offline skill handler
func (m *OfflineManager) RegisterOfflineSkill(skillID string, handler SkillHandler) {
	m.scheduler.RegisterHandler(skillID, handler)
}

// Start starts the offline manager
func (m *OfflineManager) Start() {
	m.mu.Lock()
	if m.running {
		m.mu.Unlock()
		return
	}
	m.running = true
	m.mu.Unlock()

	m.scheduler.Start()
	go m.monitorLoop()
}

// Stop stops the offline manager
func (m *OfflineManager) Stop() {
	m.mu.Lock()
	m.running = false
	m.mu.Unlock()

	m.scheduler.Stop()
	close(m.done)
	m.cache.Persist()
}

// Sync syncs pending data to center
func (m *OfflineManager) Sync(ctx context.Context, agent *Agent) error {
	// Sync pending tasks
	tasks := m.scheduler.GetSyncPendingTasks()
	for _, task := range tasks {
		if agent != nil && agent.running {
			// Send to center
			log.Printf("Syncing task %s to center", task.TaskID)
			task.SyncPending = false
		}
	}

	// Sync pending cache entries
	entries := m.cache.GetSyncPending()
	for _, entry := range entries {
		log.Printf("Syncing cache key %s to center", entry.Key)
		m.cache.MarkSynced(entry.Key)
	}

	return nil
}

func (m *OfflineManager) monitorLoop() {
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-m.done:
			return
		case <-ticker.C:
			m.autoSwitch()
			if m.IsOnline() {
				// Would sync to center here
			}
		}
	}
}

func (m *OfflineManager) autoSwitch() {
	// Auto-detect network and switch mode
	// In production: check connectivity to center
}